x/gov/handler.go

```diff
@@ -0,0 +1,76 @@
+package gov
+
+import (
+	"reflect"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// Handle all "gov" type messages.
+func NewHandler(keeper Keeper) sdk.Handler {
+	return func(ctx sdk.Context, msg sdk.Msg) sdk.Result {
+		switch msg := msg.(type) {
+		case MsgDeposit:
+			return handleMsgDeposit(ctx, keeper, msg)
+		case MsgSubmitProposal:
+			return handleMsgSubmitProposal(ctx, keeper, msg)
+		case MsgVote:
+			return handleMsgVote(ctx, keeper, msg)
+		default:
+			errMsg := "Unrecognized gov Msg type: " + reflect.TypeOf(msg).Name()
+			return sdk.ErrUnknownRequest(errMsg).Result()
+		}
+	}
+}
+
+// Handle MsgSubmitProposal.
+func handleMsgSubmitProposal(ctx sdk.Context, keeper Keeper, msg MsgSubmitProposal) sdk.Result {
+
+	proposal := keeper.NewProposal(ctx, msg.Title, msg.Description, msg.ProposalType)
+
+	err := keeper.AddDeposit(ctx, proposal.ProposalID, msg.Proposer, msg.InitialDeposit)
+	if err != nil {
+		return err.Result()
+	}
+
+	proposalIDBytes, _ := keeper.cdc.MarshalBinaryBare(proposal.ProposalID)
+
+	tags := sdk.NewTags("action", []byte("submitProposal"), "proposer", msg.Proposer.Bytes(), "proposalId", proposalIDBytes)
```

Comment:
I'm a fan of breaking out the tags to make a bit more readable: 

``` golang 
tags := sdk.NewTags(
    "action", []byte("submitProposal"), 
    "proposer", msg.Proposer.Bytes(), 
    "proposalId", proposalIDBytes,
)
```

Reply:
Additionally, all addresses need to be tagged as strings, not with `.Bytes()` - see https://github.com/cosmos/cosmos-sdk/blob/develop/x/bank/keeper.go#L154

Reply:
@cwgoes Should this not be bech32 encoded as well then?